File: services/server/src/protocol/end_packet.py
from protocol.packet import Packet, TYPE_END

END_PACKET_LEN = 1

class EndPacket(Packet):
    def __init__(self):
        super().__init__()

    def get_type(self):
        return TYPE_END

    def serialize(self):
        message = bytearray()
        length = END_PACKET_LEN.to_bytes(2, "big", signed=False)
        message.extend(length)
        message.extend(self.header())
        return message

def end_from_bytes(bytes: bytes) -> EndPacket:
    if len(bytes) != END_PACKET_LEN:
        raise ValueError("Invalid packet length")
    if bytes[0] != TYPE_END:
        raise ValueError("Invalid packet type")
    return EndPacket()

File: services/client/src/protocol/end_packet.go
package protocol

import (
	"encoding/binary"
	"errors"
)

const END_PACKET_LEN int = 1

type EndPacket struct{}

func CreateEndPacket() Packet {
	return &EndPacket{}
}

func (e *EndPacket) GetType() uint8 {
	return TYPE_END
}

func (e *EndPacket) Header() []byte {
	return []byte{e.GetType()}
}

func (e *EndPacket) Serialize() []byte {
	message := make([]byte, LENGTH_BYTES, END_PACKET_LEN+LENGTH_BYTES)
	message = binary.BigEndian.AppendUint16(message, uint16(END_PACKET_LEN))
	message = append(message, e.Header()...)
	return message
}

func EndFromBytes(bytes []byte) (*EndPacket, error) {
	if len(bytes) != END_PACKET_LEN {
		return nil, errors.New("Invalid packet length")
	}
	if bytes[0] != TYPE_END {
		return nil, errors.New("Invalid packet type")
	}
	return &EndPacket{}, nil
}
